File: packages/sdk/src/reputation/antifraud.ts
import type { Review, AntifraudFlags } from '@eterecitizen/common';

const BURST_THRESHOLD = 5; // Max reviews from same reviewer in 24h
const BURST_WINDOW_MS = 24 * 60 * 60 * 1000;
const CLUSTER_TIME_WINDOW_MS = 60 * 60 * 1000; // 1 hour
const CLUSTER_MIN_REVIEWS = 3;

export function detectAntifraudFlags(reviews: Review[]): AntifraudFlags {
  const flags: AntifraudFlags = {
    isCluster: false,
    isReciprocal: false,
    isBurst: false,
    suspicionScore: 0,
    details: [],
  };

  if (reviews.length === 0) return flags;

  // Burst detection: too many reviews from same reviewer in short time
  const reviewerTimestamps = new Map<string, Date[]>();
  for (const review of reviews) {
    const timestamps = reviewerTimestamps.get(review.reviewerDID) || [];
    timestamps.push(new Date(review.timestamp));
    reviewerTimestamps.set(review.reviewerDID, timestamps);
  }

  for (const [reviewerDID, timestamps] of reviewerTimestamps) {
    timestamps.sort((a, b) => a.getTime() - b.getTime());
    let windowStart = 0;
    for (let i = 0; i < timestamps.length; i++) {
      while (
        timestamps[i].getTime() - timestamps[windowStart].getTime() > BURST_WINDOW_MS
      ) {
        windowStart++;
      }
      if (i - windowStart + 1 >= BURST_THRESHOLD) {
        flags.isBurst = true;
        flags.suspicionScore += 0.3;
        flags.details!.push(
          `Burst: ${i - windowStart + 1} reviews from ${reviewerDID} within 24h`,
        );
        break;
      }
    }
  }

  // Cluster detection: many reviews arriving in a short time window
  const allTimestamps = reviews
    .map((r) => new Date(r.timestamp).getTime())
    .sort((a, b) => a - b);

  for (let i = 0; i <= allTimestamps.length - CLUSTER_MIN_REVIEWS; i++) {
    const windowEnd = allTimestamps[i] + CLUSTER_TIME_WINDOW_MS;
    let count = 0;
    for (let j = i; j < allTimestamps.length && allTimestamps[j] <= windowEnd; j++) {
      count++;
    }
    if (count >= CLUSTER_MIN_REVIEWS) {
      flags.isCluster = true;
      flags.suspicionScore += 0.2;
      flags.details!.push(`Cluster: ${count} reviews within 1 hour`);
      break;
    }
  }

  // Reciprocal detection: A reviewed B and B reviewed A
  const reviewPairs = new Set<string>();
  for (const review of reviews) {
    const pair = `${review.reviewerDID}->${review.reviewedDID}`;
    const reverse = `${review.reviewedDID}->${review.reviewerDID}`;
    if (reviewPairs.has(reverse)) {
      flags.isReciprocal = true;
      flags.suspicionScore += 0.4;
      flags.details!.push(
        `Reciprocal: ${review.reviewerDID} and ${review.reviewedDID} reviewed each other`,
      );
    }
    reviewPairs.add(pair);
  }

  // Cap suspicion score at 1.0
  flags.suspicionScore = Math.min(flags.suspicionScore, 1.0);

  return flags;
}

File: sdks/python/eterecitizen/antifraud.py
"""Antifraud detection (burst / cluster / reciprocal flags). See SPEC §10.5."""
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime
from typing import Final, Iterable


BURST_THRESHOLD: Final[int] = 5
BURST_WINDOW_MS: Final[int] = 24 * 60 * 60 * 1000  # 24h
CLUSTER_TIME_WINDOW_MS: Final[int] = 60 * 60 * 1000  # 1h
CLUSTER_MIN_REVIEWS: Final[int] = 3


@dataclass
class Review:
    reviewerDID: str
    reviewedDID: str
    timestamp: str          # ISO 8601
    rating: int
    category: str


@dataclass
class AntifraudFlags:
    isBurst: bool = False
    isCluster: bool = False
    isReciprocal: bool = False
    suspicionScore: float = 0.0
    details: list[str] = field(default_factory=list)


def _parse_ts_ms(iso: str) -> int:
    # Accept ISO 8601 with or without a trailing 'Z'.
    if iso.endswith("Z"):
        iso = iso[:-1] + "+00:00"
    return int(datetime.fromisoformat(iso).timestamp() * 1000)


def detect_antifraud_flags(reviews: Iterable[dict | Review]) -> AntifraudFlags:
    """Compute the three flags and cap the suspicion score at 1.0. SPEC §10.5.

    Accepts either :class:`Review` dataclasses or dict literals with the same
    keys — the conformance vectors use dicts.
    """
    flags = AntifraudFlags()

    reviews_list: list[Review] = []
    for r in reviews:
        if isinstance(r, Review):
            reviews_list.append(r)
        else:
            reviews_list.append(Review(
                reviewerDID=r["reviewerDID"],
                reviewedDID=r["reviewedDID"],
                timestamp=r["timestamp"],
                rating=r["rating"],
                category=r["category"],
            ))

    if not reviews_list:
        return flags

    # --- Burst: >= BURST_THRESHOLD reviews from same reviewer within 24h.
    by_reviewer: dict[str, list[int]] = {}
    for r in reviews_list:
        by_reviewer.setdefault(r.reviewerDID, []).append(_parse_ts_ms(r.timestamp))
    for reviewer, timestamps in by_reviewer.items():
        timestamps.sort()
        window_start = 0
        for i, ts in enumerate(timestamps):
            while ts - timestamps[window_start] > BURST_WINDOW_MS:
                window_start += 1
            if i - window_start + 1 >= BURST_THRESHOLD:
                flags.isBurst = True
                flags.suspicionScore += 0.3
                flags.details.append(
                    f"Burst: {i - window_start + 1} reviews from {reviewer} within 24h"
                )
                break

    # --- Cluster: >= CLUSTER_MIN_REVIEWS reviews across all reviewers within 1h.
    all_ts = sorted(_parse_ts_ms(r.timestamp) for r in reviews_list)
    for i in range(len(all_ts) - CLUSTER_MIN_REVIEWS + 1):
        window_end = all_ts[i] + CLUSTER_TIME_WINDOW_MS
        count = 0
        for j in range(i, len(all_ts)):
            if all_ts[j] > window_end:
                break
            count += 1
        if count >= CLUSTER_MIN_REVIEWS:
            flags.isCluster = True
            flags.suspicionScore += 0.2
            flags.details.append(f"Cluster: {count} reviews within 1 hour")
            break

    # --- Reciprocal: A→B and B→A both present.
    seen_pairs: set[str] = set()
    for r in reviews_list:
        pair = f"{r.reviewerDID}->{r.reviewedDID}"
        reverse = f"{r.reviewedDID}->{r.reviewerDID}"
        if reverse in seen_pairs:
            flags.isReciprocal = True
            flags.suspicionScore += 0.4
            flags.details.append(
                f"Reciprocal: {r.reviewerDID} and {r.reviewedDID} reviewed each other"
            )
        seen_pairs.add(pair)

    flags.suspicionScore = min(flags.suspicionScore, 1.0)
    return flags
